pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,350 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/gorilla/websocket"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	errEmptyGatewayEUI           = errors.Define("empty_gateway_eui", "empty gateway EUI")
+	errMessageTypeNotImplemented = errors.DefineUnimplemented("message_type_not_implemented", "message of type `{type}` is not implemented")
+)
+
+type srv struct {
+	ctx      context.Context
+	server   io.Server
+	upgrader *websocket.Upgrader
+	tokens   io.DownlinkTokens
+}
+
+func (*srv) Protocol() string { return "basicstation" }
+
+// New returns a new Basic Station frontend that can be registered in the web server.
+func New(ctx context.Context, server io.Server) web.Registerer {
+	ctx = log.NewContextWithField(ctx, "namespace", "gatewayserver/io/basicstation")
+	return &srv{
+		ctx:      ctx,
+		server:   server,
+		upgrader: &websocket.Upgrader{},
+	}
+}
+
+func (s *srv) RegisterRoutes(server *web.Server) {
+	group := server.Group(ttnpb.HTTPAPIPrefix + "/gs/io/basicstation")
+	group.GET("/discover", s.handleDiscover)
+	group.GET("/traffic/:id", s.handleTraffic)
+}
+
+func (s *srv) handleDiscover(c echo.Context) error {
+	ctx := c.Request().Context()
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	_, data, err := ws.ReadMessage()
+	if err != nil {
+		logger.WithError(err).Debug("Failed to read message")
+		return err
+	}
+	var req messages.DiscoverQuery
+	if err := json.Unmarshal(data, &req); err != nil {
+		logger.WithError(err).Debug("Failed to parse discover query message")
+		return err
+	}
+
+	if req.EUI.IsZero() {
+		writeDiscoverError(s.ctx, ws, "Invalid request")
```

Comment:
Is this "Invalid request" because of specification? Otherwise I would be more specific in telling why the request was invalid.

Reply:
The protocol does not define error messages nor does it process the text. It's directly logged.
These are just adapted from the example python bridge. 
https://github.com/lorabasics/basicstation/blob/c29b8502f8c715daecec6666835da6e981dc820a/src/tc.c#L221

(PS: This error would never occur in the first place, since the station itself has error handling for invalid EUIs and would not start if the config is wrong.)